hedera-node/docs/design/services/smart-contract-service/hedera-account-service-isAuthorized-design.md

```diff
@@ -0,0 +1,136 @@
+# HIP-632 (Hedera Account Service) `isAuthorized{Raw}` design
+
+This note addresses the implementation of `isAuthorized` and `isAuthorizedRaw` of 
+[HIP-632](https://hips.hedera.com/hip/hip-632) - "Hedera Account Service (HAS) 
+System Contract".
+
+## Preliminary notice w.r.t. "virtual addresses"
+
+HIP-632 as written refers a lot to 
+[HIP-631](https://hips.hedera.com/hip/hip-631) - "Account Virtual Addresses".
+However, that HIP, though `Accepted` will not be implemented.  (Or at least, not
+at any time soon.)  So HIP-632 needs to be read with that in mind.  Specifically, 
+anywhere that there's a provision for dealing with "virtual addresses", of which there
+may be several/many per Hedera account, there is in fact only _one_: the _alias_.
+
+* This mainly affects the method `getVirtualAddresses` which will return an array, but
+it will have _at most one_ EVM address in it.
+* An issue is present to edit HIP-632 to clarify that it is dealing with aliases, not virtual addresses.  
+  
+## Phased implementation
+
+The first release of HIP-632 will include only `isAuthorizedRaw`.
+
+## Where does this go?
+
+A "hedera account service" system contract is being set up in
+`com.hedera.node.app.service.contract.impl.exec.systemcontracts.has`, and also at
+`...contract.impl.exec.processors.HasTranslatorsModule` (for the individual methods).
+
+Following the existing pattern for HAS each method will be given a package under 
+`...contract.impl.exec.systemcontracts.has` with two classes:
+* `[method]Translator` - which first confirms that the call is for the given method, and
+  then builds a `TransactionBody` for it.
+* `[method]Call` - which takes that `TransactionBody` and then has an `execute()` method
+  that does the work.
+  
+## Design for `isAuthorizedRaw`
+
+`isAuthorizedRaw` has two flows: Either the argument address is a Hedera account, or it is
+an EVM address.
+
+* If an EVM address then this is simply the `ECRECOVER` precompile.  We have direct
+access to the `ECRECOVER` precompile from BESU in the class
+`org.hyperledger.besu.evm.precompile.ECRECPrecompiledContract` which will compute
+the `ECRECOVER` operation given the message hash + signature (encoded as a calldata for
+the precompile) (and a `MessageFrame` which it ignores).  This returns either the
+recovered address, if successful, or "empty", if not.  The recovered address must then
+match the address given as an argument to `isAuthorizedRaw`.
```

Comment:
It might be important to mention somewhere (perhaps the HIP is a better place for this) that the signature must be in the compressed form for `ecrecover` to succeed.  
I don't know if we want to relax this necessity in the future by implementing our own recover method verses what Besu does?

Reply:
Wait - I thought that was only when you were doing the signing that in the test code you needed the 64 byte form?

Reply:
Yes the 65 byte form but it should be constructed using the compressed form and attaching the `v` value.  I think if this is what I had to do to get the test to work, anyone trying to use the function will have to do the same?

Reply:
Not sure.  The (internal) doc that motivated this call showed examples from various sources including Uniswap v2 and v3 where their contract functions that actually called ECRECOVER had passed in `r`, s`, and `v` as separate arguments, plus another example from 0x protocol where they had a datatype `Signature` that had separate `r`, `s`, and `v` fields.  So, based on that data set, having the separate components of the signature available seems common, or, at least, familiar.